cpp/src/lists/contains.cu

```diff
@@ -44,296 +41,231 @@
-{
-  return column_device_view::create(search_keys, stream);
-}
+/**
+ * @brief A sentinel value used for marking that a given key has not been found in the search list.
+ */
+auto constexpr __device__ NOT_FOUND_SENTINEL = size_type{-1};
 
-auto get_search_keys_device_iterable_view(cudf::scalar const& search_key, rmm::cuda_stream_view)
+template <typename Element>
+auto constexpr is_supported_non_nested_type()
 {
-  return &search_key;
+  return cudf::is_fixed_width<Element>() || std::is_same_v<Element, cudf::string_view>;
 }
 
-template <typename ElementType, duplicate_find_option find_option>
-auto __device__ find_begin(list_device_view const& list)
+/**
+ * @brief Return a pair of index iterators {begin, end} to loop through elements within a list.
+ *
+ * Depending on the value of `find_first`, a pair of forward or reverse iterators will be returned,
+ * allowing to loop through elements in the list in first-to-last or last-to-first order.
+ *
+ * Note that the element indices always restart to `0` at the first position in each list.
+ *
+ * @tparam find_first A boolean value indicating whether we want to find the first or last
+ *         appearance of a given key in the list.
+ * @param size The number of elements in the list.
+ * @return A pair of {begin, end} iterators to iterate through the range `[0, size)` by forward or
+ *         reverse order.
+ */
+template <bool find_first>
+auto __device__ element_index_pair_iter(size_type const size)
 {
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return list.pair_rep_begin<ElementType>();
+  if constexpr (find_first) {
+    return thrust::pair(thrust::make_counting_iterator(0), thrust::make_counting_iterator(size));
   } else {
-    return thrust::make_reverse_iterator(list.pair_rep_end<ElementType>());
+    return thrust::pair(thrust::make_reverse_iterator(thrust::make_counting_iterator(size)),
+                        thrust::make_reverse_iterator(thrust::make_counting_iterator(0)));
   }
 }
 
-template <typename ElementType, duplicate_find_option find_option>
-auto __device__ find_end(list_device_view const& list)
-{
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return list.pair_rep_end<ElementType>();
-  } else {
-    return thrust::make_reverse_iterator(list.pair_rep_begin<ElementType>());
-  }
-}
+template <typename Element, typename SearchKeyIter>
+struct search_lists_fn {
+  cudf::detail::lists_column_device_view const lists;
+  SearchKeyIter const keys_iter;
+  duplicate_find_option const find_option;
 
-template <duplicate_find_option find_option, typename Iterator>
-size_type __device__ distance([[maybe_unused]] Iterator begin, Iterator end, Iterator find_iter)
-{
-  if (find_iter == end) {
-    return absent_index;  // Not found.
+  __device__ thrust::pair<size_type, bool> operator()(size_type const idx) const
+  {
+    auto const list = list_device_view{lists, idx};
```

Comment:
This functor takes the `idx` just to construct the `list_device_view`. Does `lists_column_device_view` have an iterator that just constructs a `list_device_view` for each of the elements in the column? 

Then this functor could accept a `list_device_view` as the input and the algorithm could be a `transform` instead of a `tabulate` (which would be more natural).

Reply:
`idx` is also used to retrieve key, as below (line 97).